### PR TITLE
Adding and removing tags

### PR DESCRIPTION
**Pull request by si-x-li**
Adding abilities to create, remove tags and assign tags to courses

### CODE REVIEW

File: backend/routes/tags.js [hunk 1]
```diff
@@ -18,4 +18,109 @@ router.get("/getAllTags", async function(req, res, next) {
...
+ * Curl:
+ *	curl -X POST
+ *	http://localhost:3001/tags/deleteTag
+ *	-H 'Content-Type: application/json'
+ *	-d '{
+ *       "tag": "Engineering"
+ *      }'
+ * @author: Steven Li
+ */
+router.post("/deleteTag", async function (req, res, next) {
```
Comment (Ty-Won) [inline backend/routes/tags.js]: Could you use the delete rest request?

File: backend/logic/tags/tags.js [hunk 2]
```diff
@@ -18,6 +25,102 @@ var getAllTags = async () => {
...
+ * Adds a tag into the database.
+ * @author Steven Li
+ * @param {String} tag
+ * @returns true if successful
+ * @throws error if MySQL connection failed
+ *         false if insertion failed
+ *         invalid format tag if tag is null
+ */
+var addTag = async tag => {
+  let connection = await mysql.getNewConnection();
```
Comment (zerfuth) [inline backend/logic/tags/tags.js]: Minor detail but I think you should get the connection after verifying the tag format

File: backend/logic/tags/tags.js [hunk 3]
```diff
@@ -18,6 +25,102 @@ var getAllTags = async () => {
...
+ * Removes a tag from the database.
+ * @author Steven Li
+ * @param {String} tag
+ * @returns true if successful
+ * @throws error if MySQL connection failed
+ *         false if deletion failed
+ *         invalid format tag if tag is null
+ */
+var removeTag = async tag => {
+  let connection = await mysql.getNewConnection();
```
Comment (zerfuth) [inline backend/logic/tags/tags.js]: same as above

File: backend/logic/tags/tags.js [hunk 4]
```diff
@@ -18,6 +25,102 @@ var getAllTags = async () => {
...
+  }
+};
+
+/**
+ * Assigns tags to a course.
+ * @param {String} course A course code
+ * @param {Array} tags An array of tags
+ */
+var assignTagsToCourse = async (course, tags) => {
+  let connection = await mysql.getNewConnection();
```
Comment (zerfuth) [inline backend/logic/tags/tags.js]: same as above

File: backend/routes/tags.js (hunk 1 above)
Comment (si-x-li) [reply backend/routes/tags.js]: Fixed

File: backend/logic/tags/tags.js (hunk 3 above)
Comment (si-x-li) [reply backend/logic/tags/tags.js]: Good catch would have caused problems otherwise

File: backend/logic/tags/tags.js [hunk 5]
```diff
@@ -18,6 +25,102 @@ var getAllTags = async () => {
...
+  await format.verifyCourseCode(course);
+  if (!tags) {
+    throw new Error("invalid format tag");
+  }
+  for (let i = 0; i < tags.length; i++) {
+    await format.verifyTag(tags[i]);
```
Comment (KevenLYF) [inline backend/logic/tags/tags.js]: no need to use wait, I think verifyTag is not an async operation

File: backend/logic/tags/tags.js [hunk 6]
```diff
@@ -18,6 +25,102 @@ var getAllTags = async () => {
...
+  try {
+    await connection.beginTransaction();
+    // Remove old entries
+    await connection.query(
+      "DELETE FROM course_tags WHERE course_code = ?;",
+      course
+    );
+
+    // Add new entries
+    for (let i = 0; i < tags.length; i++) {
```
Comment (KevenLYF) [inline backend/logic/tags/tags.js]: just a minor suggestion, you can use functional programming.
e.g
```
tags. forEach(tag => {
// do sth
})
```

File: backend/logic/tags/tags.js (hunk 5 above)
Comment (KevenLYF) [inline backend/logic/tags/tags.js]: minor suggestion, you can also do 
```
tags.forEach( tag => {
// do stuff
})

File: backend/routes/tags.js [hunk 7]
```diff
@@ -18,4 +18,109 @@ router.get("/getAllTags", async function(req, res, next) {
...
+ * @author: Steven Li
+ */
+router.post("/assignTagsToCourse", async function (req, res, next) {
+  let course = req.body.course;
+  let tag;
+
+  try {
+    tag = JSON.parse(req.body.tags);
+  } catch (err) {
+    res.status(500).send("invalid format tags");
```
Comment (KevenLYF) [inline backend/routes/tags.js]: 404

File: backend/test/tags_test.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  });
+
+  it("responds with invalid format tag on delete", async () => {
+    await tags.removeTag(null).catch(response => {
+      assert.equal(response.message, expectedInvalidFormat);
+    });
+  });
+
+  it("responds with true", async () => {
+    before(async () => {
```
Comment (KevenLYF) [inline backend/test/tags_test.js]: question, why do you put `before` and `after` in a `it` ?

File: backend/validation/format.js [hunk 9]
```diff
@@ -324,6 +324,16 @@ var verifyPhaseOut = phaseOut => {
   }
 };
 
+/**
+ * Verifies that the tag is not null
+ * @param {String} tag
+ */
+var verifyTag = async tag => {
```
Comment (KevenLYF) [inline backend/validation/format.js]: don't think this is an async function

File: backend/logic/tags/tags.js [hunk 10]
```diff
@@ -18,6 +25,102 @@ var getAllTags = async () => {
...
+/**
+ * Assigns tags to a course.
+ * @param {String} course A course code
+ * @param {Array} tags An array of tags
+ */
+var assignTagsToCourse = async (course, tags) => {
+  let connection = await mysql.getNewConnection();
+  await format.verifyCourseCode(course);
+  if (!tags) {
+    throw new Error("invalid format tag");
```
Comment (KevenLYF) [inline backend/logic/tags/tags.js]: `Invalid format tags array` ?

**Pull request merged**